source/dem/periodic_boundaries_manipulator.cc

```diff
@@ -0,0 +1,194 @@
+                                    periodic_boundary_information)});
+                }
+            }
+        }
+    }
+}
+
+template <int dim>
+void
+PeriodicBoundariesManipulator<dim>::check_and_move_particles(
+  boundary_cells_info_struct<dim> &cell_1,
+  boundary_cells_info_struct<dim> &cell_2,
+  typename Particles::ParticleHandler<dim>::particle_iterator_range
+    &particles_in_cell)
+{
+  for (auto particle = particles_in_cell.begin();
+       particle != particles_in_cell.end();
+       ++particle)
+    {
+      Point<dim, double> particle_position = particle->get_location();
+
+      // Calculate distance between particle position and the face of
+      // boundary cell d = n•(pt_particle - pt_face)
+      double distance_with_face =
+        scalar_product(particle_position - cell_1.point_on_face,
+                       cell_1.normal_vector);
+
+
+      // If distance >= 0, particle is outside of cell (or on face).
+      // If so, particle location is modified to get moved in the periodic cell.
+      if (distance_with_face >= 0.0)
+        {
+          double distance_between_faces =
+            cell_2.point_on_face[direction] - cell_1.point_on_face[direction];
+
+          // Move particle outside the current cell to the periodic cell.
+          particle_position[direction] += distance_between_faces;
+          particle->set_location(particle_position);
+        }
+    }
+}
+
+template <int dim>
+void
+PeriodicBoundariesManipulator<dim>::execute_particles_displacement(
+  const Particles::ParticleHandler<dim> &particle_handler)
+{
+  if (!periodic_boundary_cells_information.empty())
+    {
+      for (auto boundary_cells_information_iterator =
+             periodic_boundary_cells_information.begin();
+           boundary_cells_information_iterator !=
+           periodic_boundary_cells_information.end();
+           ++boundary_cells_information_iterator)
+        {
+          // Get the cell and periodic cell content from map
+          auto boundary_cells_content =
+            boundary_cells_information_iterator->second.first;
+          auto cell = boundary_cells_content.cell;
+
+          auto periodic_boundary_cells_content =
+            boundary_cells_information_iterator->second.second;
+          auto periodic_cell = periodic_boundary_cells_content.cell;
+
+          // Since cell pairs are mapped once <cell at outlet, cell at PB>,
+          // both cells in pair are checked individually, so we know the
+          // cell the particle is coming from and will execute displacement
+          // with the locally owned cell.
+          if (cell->is_locally_owned())
+            {
+              typename Particles::ParticleHandler<dim>::particle_iterator_range
+                         particles_in_cell = particle_handler.particles_in_cell(cell);
+              const bool particles_exist_in_main_cell =
+                !particles_in_cell.empty();
+
+              if (particles_exist_in_main_cell)
```

Comment:
Why not instead request the number of particles in the cell? this would improve readability instead of looking at the iterator range.

Reply:
Ok, will have to change it in another place also!